### PR TITLE
feat: Router changes to handle tenant id

### PR DESCRIPTION
**Pull request by KShivendu**
## Summary of change

Router changes to handle tenant id

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens_python/constants.py`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `setup.py`
    -   In `supertokens_python/constants.py`
-   [x] Had installed and ran the pre-commit hook
-   [x] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `supertokens_python/utils.py` file to include that in the `FRAMEWORKS` variable
-   [ ] If added a new recipe that has a User type with extra info, then be sure to change the User type in supertokens_python/types.py
-   [ ] Make sure that `syncio` / `asyncio` functions are consistent.

### CODE REVIEW

File: tests/multitenancy/test_router.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+@fixture(scope="function")
+async def client():
+    app = FastAPI()
+    app.add_middleware(get_middleware())
+
+    # @app.get("/login")
+    # async def login(_request: Request):  # type: ignore
+    #     user_id = "userId"
+    #     # await create_new_session(request, user_id, {}, {})
+    #     return {"userId": user_id}
```
Comment (rishabhpoddar) [inline tests/multitenancy/test_router.py:38]: why is this commented?

File: tests/multitenancy/test_router.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        headers={"Content-Type": "application/json"},
+        json={
+            "formFields": [
+                {"id": "password", "value": "password1"},
+                {"id": "email", "value": "test1@example.com"},
+            ]
+        },
+    )
+
+    assert res.status_code == 200  # FIXME: failing test
```
Comment (rishabhpoddar) [inline tests/multitenancy/test_router.py:104]: why this FIXme?!! Please stop making PRs with this fixme.. fix it instead

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: supertokens_python/recipe/thirdparty/providers/bitbucket.py [hunk 3]
```diff
@@ -38,9 +38,9 @@ def __init__(
         authorisation_redirect: Union[
             None, Dict[str, Union[str, Callable[[BaseRequest], str]]]
         ] = None,
-        is_default: bool = False,
+        _is_default: bool = False,
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/thirdparty/providers/bitbucket.py]: Removed. Gitlab and bitbucket haven't been implement in Node SDK yet.

File: supertokens_python/recipe/thirdparty/providers/gitlab.py [hunk 4]
```diff
@@ -41,9 +41,9 @@ def __init__(
             None, Dict[str, Union[str, Callable[[BaseRequest], str]]]
         ] = None,
         gitlab_base_url: str = "https://gitlab.com",
-        is_default: bool = False,
+        _is_default: bool = False,
     ):
-        super().__init__("gitlab") # FIXME: Where should is_default go?
+        super().__init__("gitlab")  # FIXME: Where should is_default go?
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/thirdparty/providers/gitlab.py]: Done.

File: supertokens_python/recipe_module.py [hunk 5]
```diff
@@ -42,17 +50,39 @@ def get_app_info(self):
...
+        regex = rf"^{base_path_str}(?:/([a-zA-Z0-9-]+))?(/.*)$"
+
+        match = re.match(regex, path_str)
+        tenant_id: str = DEFAULT_TENANT_ID
+        remaining_path: Optional[NormalisedURLPath] = None
+
+        if match is not None:
+            # TODO: Do something better than assert here
+            # assert match.group(1) is not None
+            # assert match.group(2) is not None
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe_module.py]: Fixed now

File: tests/multitenancy/test_router.py (hunk 1 above)
Comment (KShivendu) [reply tests/multitenancy/test_router.py]: Removed.

File: supertokens_python/recipe/dashboard/recipe.py [hunk 6]
```diff
@@ -245,15 +248,47 @@ def reset():
...
+        regex = rf"^{base_path_str}(?:/([a-zA-Z0-9-]+))?(/.*)$"
+
+        match = re.match(regex, path_str)
+        tenant_id: str = DEFAULT_TENANT_ID
+        remaining_path: Optional[NormalisedURLPath] = None
+
+        if match is not None:
+            # TODO: Do something better than assert here
+            assert match.group(1) is not None
+            assert match.group(2) is not None
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/dashboard/recipe.py]: Fixed. 

If you give:  `/dashboard` then `match.group(1)` is `None` while `match.group(2)` is `/dashboard`

File: supertokens_python/recipe/thirdparty/providers/bitbucket.py [hunk 7]
```diff
@@ -38,9 +38,9 @@ def __init__(
...
     ):
-        super().__init__("bitbucket") # FIXME: Where should is_default go?
+        super().__init__("bitbucket")  # FIXME: Where should is_default go?
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/thirdparty/providers/bitbucket.py]: removed

File: supertokens_python/recipe/thirdparty/providers/gitlab.py [hunk 8]
```diff
@@ -15,89 +15,15 @@
...
-    UserInfo,
-    UserInfoEmail,
-)
+from .custom import GenericProvider, NewProvider
+from ..provider import Provider, ProviderInput
 
-if TYPE_CHECKING:
-    from supertokens_python.framework.request import BaseRequest
 
+# TODO Implement when it's done in Node PR
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/thirdparty/providers/gitlab.py]: remove all todos from code and transfer them to https://github.com/supertokens/supertokens-python/pull/276

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: supertokens_python/recipe/thirdparty/providers/gitlab.py (hunk 8 above)
Comment (KShivendu) [reply supertokens_python/recipe/thirdparty/providers/gitlab.py]: Done.

**Pull request merged**